arangod/Aql/RemoteExecutor.cpp

```diff
@@ -0,0 +1,464 @@
+
+    if (!slice.hasKey(StaticStrings::Error) || slice.get(StaticStrings::Error).getBoolean()) {
+      THROW_ARANGO_EXCEPTION(TRI_ERROR_CLUSTER_AQL_COMMUNICATION);
+    }
+    size_t skipped = 0;
+    VPackSlice s = slice.get("skipped");
+    if (s.isNumber()) {
+      auto value = s.getNumericValue<int64_t>();
+      if (value < 0) {
+        THROW_ARANGO_EXCEPTION_MESSAGE(TRI_ERROR_BAD_PARAMETER,
+                                       "skipped cannot be negative");
+      }
+      skipped = s.getNumericValue<size_t>();
+    }
+
+    // TODO Check if we can get better with HASMORE/DONE
+    if (skipped == 0) {
+      return {ExecutionState::DONE, skipped};
+    }
+    return {ExecutionState::HASMORE, skipped};
+  }
+
+  // For every call we simply forward via HTTP
+
+  VPackBuilder builder;
+  builder.openObject();
+  builder.add("atMost", VPackValue(atMost));
+  builder.close();
+
+  auto bodyString = std::make_shared<std::string const>(builder.slice().toJson());
+
+  auto res = sendAsyncRequest(rest::RequestType::PUT, "/_api/aql/skipSome/", bodyString);
+  if (!res.ok()) {
+    THROW_ARANGO_EXCEPTION(res);
+  }
+
+  return {ExecutionState::WAITING, 0};
+}
+
+std::pair<ExecutionState, std::unique_ptr<AqlItemBlock>> ExecutionBlockImpl<RemoteExecutor>::traceGetSomeEnd(
+    ExecutionState state, std::unique_ptr<AqlItemBlock> result) {
+  ExecutionBlock::traceGetSomeEnd(result.get(), state);
+  return {state, std::move(result)};
+}
+
+std::pair<ExecutionState, size_t> ExecutionBlockImpl<RemoteExecutor>::traceSkipSomeEnd(
+    ExecutionState state, size_t skipped) {
+  ExecutionBlock::traceSkipSomeEnd(skipped, state);
+  return {state, skipped};
+}
+
+std::pair<ExecutionState, Result> ExecutionBlockImpl<RemoteExecutor>::initializeCursor(
+    AqlItemBlock* items, size_t pos) {
+  // For every call we simply forward via HTTP
+
+  if (!_isResponsibleForInitializeCursor) {
+    // do nothing...
+    return {ExecutionState::DONE, TRI_ERROR_NO_ERROR};
+  }
+
+  if (items == nullptr) {
+    // we simply ignore the initialCursor request, as the remote side
+    // will initialize the cursor lazily
+    return {ExecutionState::DONE, TRI_ERROR_NO_ERROR};
+  }
+
+  if (_lastResponse != nullptr || _lastError.fail()) {
+    // We have an open result still.
+    std::shared_ptr<VPackBuilder> responseBodyBuilder = stealResultBody();
+
+    // Result is the response which is an object containing the ErrorCode
+    VPackSlice slice = responseBodyBuilder->slice();
+    if (slice.hasKey("code")) {
+      return {ExecutionState::DONE, slice.get("code").getNumericValue<int>()};
+    }
+    return {ExecutionState::DONE, TRI_ERROR_INTERNAL};
+  }
+
+  VPackOptions options(VPackOptions::Defaults);
+  options.buildUnindexedArrays = true;
+  options.buildUnindexedObjects = true;
+
+  VPackBuilder builder(&options);
+  builder.openObject();
+
+  // Backwards Compatibility 3.3
+  builder.add("exhausted", VPackValue(false));
```

Comment:
Happy to delete this now!

Reply:
@mchacki Removing this broke the cluster tests 😕. So I re-added it, tests are running again.

Reply:
=0, i am unhappy about this.
Need to investigate who still checks exhausted.

Reply:
I'll open an issue, we'll solve this separately and merge this PR independently - agreed?